docs/proposals/17-ratelimits-for-shoot-dns-entries.md

```diff
@@ -0,0 +1,112 @@
+- Deletion of `DNSEntries` is not rate limited to avoid freezing on shoot cluster deletion.
+- Updates of `DNSEntries` created by Gardener (e.g. `internal`, `external`, and `ingress`) should not be blocked by the rate limits.
+- If creation/updates of `DNSEntries` of the `shoot-dns-service` are pending because of a rate limit, this should also made be visible in the shoot cluster (either by events or status update)
+
+### Non-Goals
+
+- prioritize updates between `DNSEntries` assigned to the same `DNSProvider`
+
+## Proposal
+
+This proposal involves changes in two components:
+
+- [dns-controller-manager](https://github.com/gardener/external-dns-management) for implementation of rate limiting
+
+- [gardener](https://github.com/gardener/gardener) for configuration of rate limits, applying them to the external `DNSProvider` and annotating `DNSEntries` to exclude then from rate limiting
+
+### DNS controller manager
+
+The spec of a `DNSProvider` is extended by an optional `rateLimit` section.
+
+```go
+type DNSProviderSpec struct {
+    ...
+    // rate limit for create/update operations on DNSEntries assigned to this provider
+    // +optional
+    RateLimit *RateLimit `json:"rateLimit,omitempty"`
+}
+
+type RateLimit struct {
+    // RequestsPerDay is create/update request rate per DNS entry given by requests per day
+    RequestsPerDay int `json:"requestsPerDay"`
+    // Burst allows bursts of up to 'burst' to exceed the rate defined by 'RequestsPerDay', while still maintaining a
+    // smoothed rate of 'RequestsPerDay'
+    Burst int `json:"burst"`
+}
+```
+
+Example manifest:
+
+```yaml
+kind: DNSProvider
+...
+spec:
+  ...
+  rateLimit:
+    requestsPerDay: 120
+    burst: 20
+```
+
+Each valid `DNSEntry` is already automatically assigned to an `DNSProvider` by matching the domain name and the realm.
+This assignment now also includes the regime of the rate limit of this `DNSProvider`.
+The creation or update of DNS records is postponed until the rate limit restrictions are fulfilled.
+The entry stays in the state `Pending` or `Updating` until the request is applied on the DNS cloud service.
+Here the state `Updating` is a new, additional state to be implemented to make transparent that spec and backend are not yet synchronized.
+
+Deletion of `DNSEntry` is always applied immediately to avoid possible extreme delays on shoot deletion.
+
+As the status of `DNSEntries` created by the `shoot-dns-service` in the CP is already forwarded to the source objects in the shoot cluster in form
+of events, nothing needs to be added here. Additionally, the status for shoot-side `DNSEntries` is updated accordingly.
+
+On replicating a `DNSEntry` from the shoot cluster nothing must be changed, as annotations are not copied. This means a user cannot escape the rate limiting by adding an annotation like `dns.gardener.cloud/not-rate-limited: "true"`
+
+### Gardener
```

Comment:
How is this rate limit for the default provider determined? 
Should that be defined by an operator during the Gardener installation e.g together with the default domain secret?

Just an idea: what about setting the maximum daily requests possible for a certain provider (and region?) and Gardener determines a "fair share" of requests per Shoot cluster. 
This limit could grow and shrink based on the number of Shoot clusters.
However, it is possibly too complex to implement & no benefit besides having too low/high rate limits.

Reply:
We have collected some statistics from samples over one week on the canary and live landscapes. The maximum seen number of DNS entries created by the shoot-dns-service within one hour was 10. This number only contains DNS entries in the default domain.
At least with the current usage pattern of our users, this should be a good starting point (i.e. a rate of 240 creations/updates per day).